Index: SpecialUndelete.php
===================================================================
--- SpecialUndelete.php	(revision 2710)
+++ SpecialUndelete.php	(working copy)
@@ -806,7 +806,28 @@
 			} else {
 				$this->showFile( $this->mFilename );
 			}
-		} elseif ( $this->mRestore && $this->mAction == 'submit' ) {
+		}
+		// CORE HACK for [[mw:Extension:Video]]
+		elseif ( $this->mTargetObj->inNamespace( NS_VIDEO ) ) {
+			$file = new ArchivedVideo( $this->mTargetObj, '', $this->mFilename );
+			// Check if user is allowed to see this file
+			if ( !$file->exists() ) {
+				$out->addWikiMsg( 'filedelete-nofile', $this->mFilename );
+			} elseif ( !$file->userCan( File::DELETED_FILE, $user ) ) {
+				if ( $file->isDeleted( File::DELETED_RESTRICTED ) ) {
+					throw new PermissionsError( 'suppressrevision' );
+				} else {
+					throw new PermissionsError( 'deletedtext' );
+				}
+			}
+			if ( $this->mRestore && $this->mAction == 'submit' ) {
+				$this->undelete();
+			} else {
+				$this->showHistory();
+			}
+		}
+		// END CORE HACK
+		elseif ( $this->mRestore && $this->mAction == 'submit' ) {
 			$this->undelete();
 		} else {
 			$this->showHistory();
@@ -1240,8 +1261,15 @@
 		if ( $haveFiles ) {
 			$batch = new LinkBatch();
 			foreach ( $files as $row ) {
-				$batch->addObj( Title::makeTitleSafe( NS_USER, $row->fa_user_text ) );
-				$batch->addObj( Title::makeTitleSafe( NS_USER_TALK, $row->fa_user_text ) );
+				// CORE HACK for [[mw:Extension:Video]]
+				if ( isset( $row->ov_user_name ) && $row->ov_user_name ) {
+					$batch->addObj( Title::makeTitleSafe( NS_USER, $row->ov_user_name ) );
+					$batch->addObj( Title::makeTitleSafe( NS_USER_TALK, $row->ov_user_name ) );
+				} else {
+					$batch->addObj( Title::makeTitleSafe( NS_USER, $row->fa_user_text ) );
+					$batch->addObj( Title::makeTitleSafe( NS_USER_TALK, $row->fa_user_text ) );
+				}
+				// END CORE HACK
 			}
 			$batch->execute();
 			$files->seek( 0 );
@@ -1458,6 +1486,27 @@
 	}
 
 	private function formatFileRow( $row ) {
+		// CORE HACK for [[mw:Extension:Video]]
+		if ( isset( $row->ov_name ) && $row->ov_name ) {
+			$file = ArchivedVideo::newFromRow( $row );
+			$ts = wfTimestamp( TS_MW, $row->ov_timestamp );
+			$user = $this->getUser();
+
+			if ( $this->mAllowed ) { //&& $row->fa_storage_key ) {
+				$checkBox = Xml::check( 'fileid' ); //. $row->fa_id );
+				$key = rand();//urlencode( $row->fa_storage_key );
+				$pageLink = $this->getFileLink( $file, $this->getPageTitle(), $ts, $key );
+			} else {
+				$checkBox = '';
+				$pageLink = $this->getLanguage()->userTimeAndDate( $ts, $user );
+			}
+			$userLink = $this->getFileUser( $file );
+			$data = $comment = $revdlink = '';
+
+			return "<li>$checkBox $revdlink $pageLink . . ({$row->ov_type}) . . $userLink $data $comment</li>\n";
+		}
+		// END CORE HACK
+
 		$file = ArchivedFile::newFromRow( $row );
 		$ts = wfTimestamp( TS_MW, $row->fa_timestamp );
 		$user = $this->getUser();
@@ -1551,16 +1600,25 @@
 			return '<span class="history-deleted">' . $time . '</span>';
 		}
 
-		$link = Linker::linkKnown(
-			$titleObj,
-			htmlspecialchars( $time ),
-			array(),
-			array(
-				'target' => $this->mTargetObj->getPrefixedText(),
-				'file' => $key,
-				'token' => $user->getEditToken( $key )
-			)
-		);
+		// CORE HACK for [[mw:Extension:Video]]
+		if ( $file instanceof ArchivedVideo ) {
+			$link = Linker::makeExternalLink(
+				$file->getURL(), $time, /* $escape */ true, /* $linktype */ '',
+				/* $attribs */ array(), $titleObj
+			);
+		} else {
+			$link = Linker::linkKnown(
+				$titleObj,
+				htmlspecialchars( $time ),
+				array(),
+				array(
+					'target' => $this->mTargetObj->getPrefixedText(),
+					'file' => $key,
+					'token' => $user->getEditToken( $key )
+				)
+			);
+		}
+		// END CORE HACK
 
 		if ( $file->isDeleted( File::DELETED_FILE ) ) {
 			$link = '<span class="history-deleted">' . $link . '</span>';
